crates/ra_hir/src/ty/tests.rs

```diff
@@ -2228,6 +2255,61 @@ fn infer(content: &str) -> String {
     let source_file = db.parse(file_id);
     let mut acc = String::new();
     acc.push_str("\n");
+
+    for const_def in source_file.syntax().descendants().filter_map(ast::ConstDef::cast) {
+        let konst = source_binder::const_from_source(&db, file_id, const_def).unwrap();
+        let inference_result = konst.infer(&db);
+        let body_source_map = konst.body_source_map(&db);
```

Comment:
The loop body for consts, statics, and fns is mostly the same, except that fns print out inference results for patterns and expressions (`type_of_pat`, `type_of_expr`) while consts and statics only print out the latter. Is there a reason for that?

Also, since the body is mostly the same, you could factor out everything from `let mut types = Vec::new();` to the end of the loop body into a closure that takes the `inference_result` and `body_source_map` and call that from each of the three loops.

Reply:
I tried refractoring to using a closure and that causes use to lose some typing and as a result around 56 tests start failing